api/migrations/2020-11-08-174607_trigger_users/up.sql

```diff
@@ -0,0 +1,35 @@
+CREATE OR REPLACE FUNCTION delete_dangling_program_records()
+    RETURNS trigger AS
+$$
+BEGIN
+    WITH pbs as (
+        SELECT min(cpu_cycles) as cpu_cycles, min(instructions) as instructions, min(registers_used) as registers_used, min(stacks_used) as stacks_used
+        FROM "user_program_records" as upr
+        WHERE upr.user_id = NEW.user_id 
+        AND upr.program_spec_id = NEW.program_spec_id
+    )
+    DELETE FROM "user_program_records" WHERE id in (
+        SELECT id
+        FROM "user_program_records" as upr, pbs
+        WHERE upr.user_id = NEW.user_id AND upr.program_spec_id = NEW.program_spec_id
+        AND upr.cpu_cycles > pbs.cpu_cycles
+        AND upr.instructions > pbs.instructions
+        AND upr.registers_used > pbs.registers_used
+        AND upr.stacks_used > pbs.stacks_used
```

Comment:
Do you think we should index these cols? Can you check if adding the indexes lets PG not do a scan? I would think pg is smart enough for that, since the rest of the WHERE clause is only on the PK col

Reply:
Yea ill check if we have them and add them if not

Reply:
So from the looks of it indexs where you do a `>` or `<` check are not used often. I think if we just put an index on `user_id` `program_spec_id` will be good enough so it would get all records for the current user's program spec then just scan the couple records in there

Reply:
https://www.postgresql.org/docs/12/indexes-multicolumn.html